inline.h

```diff
@@ -57,6 +57,94 @@ Perl_av_count(pTHX_ AV *av)
     return AvFILL(av) + 1;
 }
 
+/* ------------------------------- av.c ------------------------------- */
+
+/*
+=for apidoc av_store_simple
+
+This is a cut-down version of av_store that assumes that the array is
+very straightforward - no magic, not readonly, and AvREAL - and that
+C<key> is not negative. This function MUST NOT be used in situations
+where any of those assumptions may not hold.
+
+Stores an SV in an array.  The array index is specified as C<key>.  The
+return value will be C<NULL> if the operation failed or if the value did not
+need to be actually stored within the array (as in the case of tied
+arrays).  Otherwise, it can be dereferenced
+to get the C<SV*> that was stored
+there (= C<val>)).
+
+Note that the caller is responsible for suitably incrementing the reference
+count of C<val> before the call, and decrementing it if the function
+returned C<NULL>.
+
+Approximate Perl equivalent: C<splice(@myarray, $key, 1, $val)>.
+
+=cut
+*/
+
+PERL_STATIC_INLINE SV**
+Perl_av_store_simple(pTHX_ AV *av, SSize_t key, SV *val)
+{
+    SV** ary;
+
+    PERL_ARGS_ASSERT_AV_STORE_SIMPLE;
+    assert(SvTYPE(av) == SVt_PVAV);
+    assert(!SvMAGICAL(av));
+    assert(!SvREADONLY(av));
+    assert(AvREAL(av));
+    assert(key > -1);
+
+    ary = AvARRAY(av);
+
+    if (AvFILLp(av) < key) {
+        if (key > AvMAX(av)) {
+            av_extend(av,key);
+            ary = AvARRAY(av);
+        }
+        AvFILLp(av) = key;
+    } else
+        SvREFCNT_dec(ary[key]);
+
+    ary[key] = val;
+    return &ary[key];
+}
+
+/*
+=for apidoc av_fetch_simple
+
+This is a cut-down version of av_fetch that assumes that the array is
+very straightforward - no magic, not readonly, and AvREAL - and that
+C<key> is not negative. This function MUST NOT be used in situations
+where any of those assumptions may not hold.
+
+Returns the SV at the specified index in the array.  The C<key> is the
+index.  If lval is true, you are guaranteed to get a real SV back (in case
+it wasn't real before), which you can then modify.  Check that the return
+value is non-null before dereferencing it to a C<SV*>.
+
+The rough perl equivalent is C<$myarray[$key]>.
+
+=cut
+*/
+
+PERL_STATIC_INLINE SV**
+Perl_av_fetch_simple(pTHX_ AV *av, SSize_t key, I32 lval)
+{
+    PERL_ARGS_ASSERT_AV_FETCH_SIMPLE;
+    assert(SvTYPE(av) == SVt_PVAV);
+    assert(!SvMAGICAL(av));
+    assert(!SvREADONLY(av));
+    assert(AvREAL(av));
+    assert(key > -1);
+
+    if ( (key >= AvFILLp(av) + 1) || !AvARRAY(av)[key]) {
```

Comment:
Why the +1?
```
  if (key > AvFILLp(av) || !AvARRAY(av)[key]) {
```

(I think this could actually be a bug in rare circumstances, if fill hits `SSize_t_MAX`.)

Reply:
It came from me refactoring the following, when we know that `neg` is untrue:
```
    neg  = (key < 0);
    size = AvFILLp(av) + 1;
    key += neg * size; /* handle negative index without using branch */

    /* the cast from SSize_t to Size_t allows both (key < 0) and (key >= size)
     * to be tested as a single condition */
    if ((Size_t)key >= (Size_t)size) {
        if (UNLIKELY(neg))
            return NULL;
        goto emptyness;
    }

    if (!AvARRAY(av)[key]) {
      emptyness:
        return lval ? av_store(av,key,newSV(0)) : NULL;
    }
```
I removed the cast, thinking that it wasn't necessary, but I now think you're right about things going astray if AvFILLp == SSize_t_MAX. (Nice spot.)

So then it should be as you suggest:

`if (key > AvFILLp(av) || !AvARRAY(av)[key]) {` ?

Or since the function assumes `AvREAL(av)`, would it actually be enough just to say:

`if (!AvARRAY(av)[key]) {` ?

Reply:
> Or since the function assumes `AvREAL(av)`, would it actually be enough just to say:
> 
> `if (!AvARRAY(av)[key]) {` ?

I don't think so, you'd be reading past the bounds of the array.

Reply:
Derp, yes. Brain has disengaged for the night.

Reply:
I'll change it to `if (key > AvFILLp(av) || !AvARRAY(av)[key]) { `

Reply:
Update pushed.